aptos-move/framework/aptos-framework/sources/voting.spec.move

```diff
@@ -0,0 +1,195 @@
+        // The next_proposal_id in VotingForum is up to MAX_U64.
+        aborts_if voting_forum.next_proposal_id + 1 > MAX_U64;
+        let is_multi_step_in_execution_key = std::string::spec_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        aborts_if is_multi_step_proposal && simple_map::spec_contains_key(metadata,is_multi_step_in_execution_key);
+    }
+
+    spec vote<ProposalType: store>(
+        _proof: &ProposalType,
+        voting_forum_address: address,
+        proposal_id: u64,
+        num_votes: u64,
+        should_pass: bool,
+    ) {
+        aborts_if !exists<VotingForum<ProposalType>>(voting_forum_address);
+        let voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::spec_get(voting_forum.proposals, proposal_id);
+        // Getting proposal from voting forum might fail because of non-exist id
+        aborts_if !table::spec_contains(voting_forum.proposals, proposal_id);
+        // Aborts when voting period is over or resolved
+        aborts_if is_voting_period_over(proposal);
+        aborts_if proposal.is_resolved;
+        // Assert this proposal is single-step, or if the proposal is multi-step, it is not in execution yet.
+        aborts_if !std::string::spec_internal_check_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        let execution_key = std::string::spec_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        aborts_if simple_map::spec_contains_key(proposal.metadata, execution_key) &&
+                  simple_map::spec_get(proposal.metadata, execution_key) != std::bcs::spec_to_bytes(false);
+        aborts_if if (should_pass) { proposal.yes_votes + num_votes > MAX_U128 } else { proposal.no_votes + num_votes > MAX_U128 };
+
+        aborts_if !std::string::spec_internal_check_utf8(RESOLVABLE_TIME_METADATA_KEY);
+    }
+
+    spec fun spec_is_proposal_resolvable(voting_forum_address: address, proposal_id: u64): bool;
+
+    spec is_proposal_resolvable<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+    ) {
+        // This function aborts when the proposal is not resolvable.
+
+        // TODO: Find a way to specify when it will abort. The opaque pragma with spec fun doesn't work.
+        pragma aborts_if_is_partial;
+        // pragma opaque;
+        // aborts_if [abstract] spec_is_proposal_resolvable(voting_forum_address, proposal_id);
+    }
+
+    spec resolve<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+    ): ProposalType {
+        pragma aborts_if_is_partial;
+        include ContainsProposalID<ProposalType>;
+        aborts_if !std::string::spec_internal_check_utf8(IS_MULTI_STEP_PROPOSAL_KEY);
+    }
+
+    spec resolve_proposal_v2<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+        next_execution_hash: vector<u8>,
+    ) {
+        pragma aborts_if_is_partial;
```

Comment:
`include IsProposalResolvableAbortsIf;` instead of making it "partial"?

Reply:
What kind of `aborts_if` specs make it difficult to be complete (i.e., not "partial")?